src/phoenix/db/bulk_inserter.py

```diff
@@ -0,0 +1,177 @@
+import asyncio
+import logging
+from itertools import islice
+from time import time
+from typing import Any, AsyncContextManager, Callable, Iterable, List, Optional, Tuple, cast
+
+from openinference.semconv.trace import SpanAttributes
+from sqlalchemy import func, insert, select, update
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from phoenix.db import models
+from phoenix.trace.schemas import Span, SpanStatusCode
+
+logger = logging.getLogger(__name__)
+
+
+class SpansBulkInserter:
+    def __init__(
+        self,
+        db: Callable[[], AsyncContextManager[AsyncSession]],
+        initial_batch: Optional[Iterable[Tuple[Span, str]]] = None,
+        sleep_seconds: float = 0.5,
+        max_num_spans_per_transaction: int = 100,
+    ) -> None:
+        self._db = db
+        self._running = False
+        self._sleep_seconds = sleep_seconds
+        self._max_num_spans_per_transaction = max_num_spans_per_transaction
+        self._batch: List[Tuple[Span, str]] = [] if initial_batch is None else list(initial_batch)
+        self._task: Optional[asyncio.Task[None]] = None
+
+    async def __aenter__(self) -> Callable[[Span, str], None]:
+        self._running = True
+        self._task = asyncio.create_task(self._insert_spans())
+        return self._queue_span
+
+    async def __aexit__(self, *args: Any) -> None:
+        self._running = False
+
+    def _queue_span(self, span: Span, project_name: str) -> None:
+        self._batch.append((span, project_name))
+
+    async def _insert_spans(self) -> None:
+        next_run_at = time() + self._sleep_seconds
+        while self._batch or self._running:
+            await asyncio.sleep(next_run_at - time())
+            next_run_at = time() + self._sleep_seconds
+            if not self._batch:
+                continue
+            batch = self._batch
+            self._batch = []
+            for i in range(0, len(batch), self._max_num_spans_per_transaction):
+                try:
+                    async with self._db() as session:
+                        for span, project_name in islice(
+                            batch, i, i + self._max_num_spans_per_transaction
+                        ):
+                            try:
+                                async with session.begin_nested():
+                                    await _insert_span(session, span, project_name)
+                            except Exception:
+                                logger.exception(
+                                    f"Failed to insert span with span_id={span.context.span_id}"
+                                )
+                except Exception:
+                    logger.exception("Failed to insert spans")
+
+
+async def _insert_span(session: AsyncSession, span: Span, project_name: str) -> None:
+    if await session.scalar(select(1).where(models.Span.span_id == span.context.span_id)):
+        # Span already exists
+        return
```

Comment:
Yes, but it'll raise a IntegrityError which is annoying. On the other hand, this operation here is not expensive, because the B-tree is most likely already in the buffer pool.